PFW-1510: Improve power panic when print is already saved in RAM

gudnimg opened the pull request:
Change the power panic code such that it will also work when a print is paused or the printer is sitting in an MMU error screen. Basically all scenarios which have called `stop_and_save_print_to_ram()` and have parked the extruder. The parking move must not overwrite what has already been written to Flash/RAM.

The following scenarios now recover the print job after a power outage (previously the print would be lost or the printer would try printing in mid-air):
* Power outage during a paused print
* Power outage during an MMU error
* Power outage during filament runout event (when waiting in M600)
    * Both FINDA and Fsensor runouts
 * Power outage during the fullscreen prompt after a crash detection (the screen which asks if you want to resume the print)


Fixes https://github.com/prusa3d/Prusa-Firmware/issues/3241
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/2325
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/2249
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/2013
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/2001
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/1859
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/1765
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/1695
Fixes https://github.com/prusa3d/Prusa-Firmware/issues/905

Change in memory:
Flash: -612 bytes
SRAM: -4 bytes

gudnimg committed:
power panic: Move code into a separate file

gudnimg committed:
power panic: Add two functions

These functions should be able to be re-used during a power panic

- save_print_file_state
- restore_print_file_state

No functional change at the moment.

gudnimg committed:
power panic: Re-use saving print file state

This also adds the Gcode position into EEPROM
for Octoprint/PrusaLink prints.
Though the recovery part doesn't support that yet.

gudnimg committed:
power panic: Re-use code to save feedrate, segment index, and start position

Saves ~80B of Flash

gudnimg committed:
power panic: Add comments about conserving power

gudnimg commented on Firmware/Marlin.h:
```diff
@@ -380,23 +382,25 @@ float temp_compensation_pinda_thermistor_offset(float temperature_pinda);
 extern void print_physical_coordinates();
 extern void print_mesh_bed_leveling_table();
 
+void save_print_file_state();
+void restore_print_file_state();
+void save_planner_global_state();
 extern void stop_and_save_print_to_ram(float z_move, float e_move);
 void restore_extruder_temperature_from_ram();
 extern void restore_print_from_ram_and_continue(float e_move);
 extern void cancel_saved_printing();
```
I wonder if we should create a separate file for these functions as well or rename `power_panic.cpp` to something else to fit this code? 🤔

gudnimg committed:
power panic: re-use target temps from RAM

If a print has been saved to RAM such as during a filament runout,
do not overwrite these saved values if a sudden
power panic appears.

Additionally, change the saved types to be the same as power panic when saving to RAM:
- Bed target temperature is uint8_t (0 to 255) instead of float
- Extruder target temperature is uint16_t instead of float

Doing this change allows us to re-use the same global variables and
avoid creating local variables during power panic.

Change in memory:
Flash: -246 bytes
SRAM: -5 bytes

gudnimg committed:
power panic: re-use saved_extruder_relative_mode

Change in memory:
Flash: +8 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: re-use saved_pos

This allows us to restore the position of all axis saved in RAM

If the extruder had been parked to the side for example
due to filament runout. Then the original position (before parking)
should now be restored

Change in memory:
Flash: +40 bytes
SRAM: 0 bytes

gudnimg commented:
The PR is working well to resolve the issue with paused prints. Now my printer recovers the print :) 

Only thing left is to run a lot of tests to make sure everything which was working before still works.

The following scenarios should now recover a print after a power outage (there may be more):

- [x] Power outage during a paused print
- [x] Power outage during an MMU error
- [x] Power outage during filament runout event (when waiting in M600)
    - Both FINDA and Fsensor runouts

gudnimg commented:
Marking as ready for review. I'll try to spend the rest of my time this weekend on testing and verification. Leaving the 🚧 in the title until testing is done.

gudnimg committed:
optimisation: read/write EEPROM_UVLO_SAVED_START_POSITION block

I verified via printf that sizeof(saved_start_position) = 16 (i.e. 4 float values)
We simply want to write 16 bytes to address
EEPROM_UVLO_SAVED_START_POSITION

Change in memory:
Flash: -160 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: cleanup: remove commented code

gudnimg committed:
power panic: store saved_fan_speed in EEPROM

fanSpeed may have an outdated value

gudnimg committed:
power panic: pull INT4 string into PROGMEM

Change in memory:
Flash: -4 bytes
SRAM 0 bytes

gudnimg committed:
power panic: use saved position consistently

1) current_position[Z_AXIS] is not always correct. saved_pos[Z_AXIS]
    should always represent the correct resume position for the Z-axis
2) Use the saved position to fetch the Z-offset value from
    the mesh bed leveling grid. In case the extruder is parked during
    power panic, the previous code may extract the wrong mesh
    bed leveling offset (due to extruder being located at different
    X and Y axis coordinates)

gudnimg committed:
power panic: implement partial back up in RAM

A partial backup is needed in scenarios where the extruder may be
parked after a print is saved. For example during a blocking wait for the user in M600
Or during a MMU error screen.

A sudden power panic at this point would previously save the parked position
into EEPROM. When the print is recovered it would print in mid air.

gudnimg commented on Firmware/power_panic.cpp:
```diff
@@ -0,0 +1,490 @@
+    if(mbl_was_active) {
+        // Mesh bed leveling was being actively applied to the Z-position. Revert the
+        // mesh bed leveling offset value.
+        logical_z -= mbl.get_z(st_get_position_mm(X_AXIS), st_get_position_mm(Y_AXIS));
```
I suspect this should be `mbl.get_z(saved_pos[X_AXIS], saved_pos[Y_AXIS])` since if the extruder had been parked prior to the power panic, the X and Y machine positions are not correct (i.e. they are not equal to the saved print position)

3d-gussner commented:
Tested on real MK3S+MMU
M155 S1 C5
1. SD print :heavy_check_mark: 
1.a. feedrate 100% :heavy_check_mark:
1.b. feedrate 90% :heavy_check_mark:
1.c. wait hotend/bed cool down to confirm :heavy_check_mark:
2. SD print paused :heavy_check_mark:
3. SD print M600 :heavy_check_mark:
4. MMU SD print :heavy_check_mark:
5. MMU SD print paused :heavy_check_mark:
6. MMU SD print during tool change :heavy_check_mark:
7. MMU SD print Error screen hot :heavy_check_mark:
8. MMU SD print Error screen cold / timed-out :heavy_check_mark:

Normally I would approve this but I want to wait for Prusa testing approval.

gudnimg committed:
fix build

gudnimg committed:
power panic: Fix an error in saved position

planner_abort_hard() calls planner_reset_position() which
will set current_position vector to the machine position.
We want to save this position when there is no position already saved
(i.e. when there is no partial back-up or a saved print in RAM)

When a power outage comes, the printer is in the middle of a gcode move.
And at the moment a gcode is executed by the planner, the planner will update
current_position vector to the final destination vector. This means current_position
vector is invalidated during a power outage and so we must check what the
actual machine position is instead and save it.

This was working correctly before, this commit only fixes the regression
in my pull request.

Change in memory:
Flash: -2 bytes
SRAM: 0 bytes

gudnimg commented:
@DRracer @3d-gussner [5d19198](https://github.com/prusa3d/Prusa-Firmware/pull/4243/commits/5d191986081cd1cac9a1a8aa3975257933302d9f) fixes the "extrusion" issue :) It was a regression in my PR and was working fine before.

gudnimg committed:
power panic: refactor log output on recovery

Use M114 to print coordinates for all axis. We save
a lot of memory by reusing the M114 code

Change in memory:
Flash: -184 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: remove unused EEPROM_UVLO_Z_MICROSTEPS

It isn't very helpful IMO to have unused code during power panic
it just complicates things.

Change in memory:
Flash: -32 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: add readable flags and optimise ISR(INT4_vect)

Add more readable names for the EEPROM_UVLO values

Change in memory:
Flash: -14 bytes
SRAM: 0 bytes

gudnimg committed:
power panic: optimise incrementing EEPROM statistics

Change in memory:
Flash: -26 bytes
SRAM: 0 bytes

DRracer commented on Firmware/power_panic.cpp:
```diff
@@ -163,10 +163,8 @@ void uvlo_() {
+    // Write the _final_ Z position
     eeprom_update_float((float*)EEPROM_UVLO_TINY_CURRENT_POSITION_Z, current_position[Z_AXIS]);
-    z_microsteps = tmc2130_rd_MSCNT(Z_AXIS);
-    eeprom_update_word((uint16_t*)(EEPROM_UVLO_Z_MICROSTEPS), z_microsteps);
```
Are you sure this part was never used? In the past, there were issues with PP not recovering the Z-height correctly in some cases. @wavexx spent a lot of time debugging that.

gudnimg replied:
Yes at least not by the firmware.  We are not reading from this EEPROM address anywhere. It was only being written to.
